crypto/engine/eng_devcrypto.c

```diff
@@ -53,79 +73,136 @@ struct cipher_ctx {
+    { NID_aes_128_xts, 16, 128 / 8 * 2, 16, EVP_CIPH_XTS_MODE, CRYPTO_AES_XTS,
+      DEVCRYPTO_STATUS_UNKNOWN, DEVCRYPTO_ACCELERATION_UNKNOWN, 1 },
+    { NID_aes_256_xts, 16, 256 / 8 * 2, 16, EVP_CIPH_XTS_MODE, CRYPTO_AES_XTS,
+      DEVCRYPTO_STATUS_UNKNOWN, DEVCRYPTO_ACCELERATION_UNKNOWN, 1 },
 #endif
 #if !defined(CHECK_BSD_STYLE_MACROS) || defined(CRYPTO_AES_ECB)
-    { NID_aes_128_ecb, 16, 128 / 8, 16, EVP_CIPH_ECB_MODE, CRYPTO_AES_ECB },
-    { NID_aes_192_ecb, 16, 192 / 8, 16, EVP_CIPH_ECB_MODE, CRYPTO_AES_ECB },
-    { NID_aes_256_ecb, 16, 256 / 8, 16, EVP_CIPH_ECB_MODE, CRYPTO_AES_ECB },
+    { NID_aes_128_ecb, 16, 128 / 8, 16, EVP_CIPH_ECB_MODE, CRYPTO_AES_ECB,
+      DEVCRYPTO_STATUS_UNKNOWN, DEVCRYPTO_ACCELERATION_UNKNOWN, 1 },
+    { NID_aes_192_ecb, 16, 192 / 8, 16, EVP_CIPH_ECB_MODE, CRYPTO_AES_ECB,
+      DEVCRYPTO_STATUS_UNKNOWN, DEVCRYPTO_ACCELERATION_UNKNOWN, 1 },
+    { NID_aes_256_ecb, 16, 256 / 8, 16, EVP_CIPH_ECB_MODE, CRYPTO_AES_ECB,
+      DEVCRYPTO_STATUS_UNKNOWN, DEVCRYPTO_ACCELERATION_UNKNOWN, 1 },
 #endif
 #if 0                            /* Not yet supported */
-    { NID_aes_128_gcm, 16, 128 / 8, 16, EVP_CIPH_GCM_MODE, CRYPTO_AES_GCM },
-    { NID_aes_192_gcm, 16, 192 / 8, 16, EVP_CIPH_GCM_MODE, CRYPTO_AES_GCM },
-    { NID_aes_256_gcm, 16, 256 / 8, 16, EVP_CIPH_GCM_MODE, CRYPTO_AES_GCM },
+    { NID_aes_128_gcm, 16, 128 / 8, 16, EVP_CIPH_GCM_MODE, CRYPTO_AES_GCM,
+      DEVCRYPTO_STATUS_UNKNOWN, DEVCRYPTO_ACCELERATION_UNKNOWN, 1 },
+    { NID_aes_192_gcm, 16, 192 / 8, 16, EVP_CIPH_GCM_MODE, CRYPTO_AES_GCM,
+      DEVCRYPTO_STATUS_UNKNOWN, DEVCRYPTO_ACCELERATION_UNKNOWN, 1 },
+    { NID_aes_256_gcm, 16, 256 / 8, 16, EVP_CIPH_GCM_MODE, CRYPTO_AES_GCM,
+      DEVCRYPTO_STATUS_UNKNOWN, DEVCRYPTO_ACCELERATION_UNKNOWN, 1 },
 #endif
 #ifndef OPENSSL_NO_CAMELLIA
     { NID_camellia_128_cbc, 16, 128 / 8, 16, EVP_CIPH_CBC_MODE,
-      CRYPTO_CAMELLIA_CBC },
+      CRYPTO_CAMELLIA_CBC,
+      DEVCRYPTO_STATUS_UNKNOWN, DEVCRYPTO_ACCELERATION_UNKNOWN, 1 },
     { NID_camellia_192_cbc, 16, 192 / 8, 16, EVP_CIPH_CBC_MODE,
-      CRYPTO_CAMELLIA_CBC },
+      CRYPTO_CAMELLIA_CBC,
+      DEVCRYPTO_STATUS_UNKNOWN, DEVCRYPTO_ACCELERATION_UNKNOWN, 1 },
     { NID_camellia_256_cbc, 16, 256 / 8, 16, EVP_CIPH_CBC_MODE,
-      CRYPTO_CAMELLIA_CBC },
+      CRYPTO_CAMELLIA_CBC,
+      DEVCRYPTO_STATUS_UNKNOWN, DEVCRYPTO_ACCELERATION_UNKNOWN, 1 },
 #endif
 };
 
-static size_t get_cipher_data_index(int nid)
+static int get_cipher_data_index(int nid)
 {
     size_t i;
 
     for (i = 0; i < OSSL_NELEM(cipher_data); i++)
         if (nid == cipher_data[i].nid)
-            return i;
-
-    /*
-     * Code further down must make sure that only NIDs in the table above
-     * are used.  If any other NID reaches this function, there's a grave
-     * coding error further down.
-     */
-    assert("Code that never should be reached" == NULL);
+            return (int) i;
     return -1;
 }
```

Comment:
Were the changes in this function actually needed?  The condition that the nid must be a known one is still true, isn't it?

Reply:
I changed the semantics of the return value.  I'm returning -1 if the NID is not found.  This function is now used to find the cipher/digest set by the user in the configuration command, so it may return an unknown cipher/digest.

Reply:
Ah, yeah I see what you're doing there...  would you indulge me in having this instead?

``` C
static size_t find_cipher_data_index(int nid)
{
    size_t i;

    for (i = 0; i < OSSL_NELEM(cipher_data); i++)
        if (nid == cipher_data[i].nid)
            return i;
    return (size_t)-1;
}

static size_t get_cipher_data_index(int nid)
{
    size_t i = find_cipher_data_index(nid);

    if (i != (size_t)-1)
        return i;

    /*
     * Code further down must make sure that only NIDs in the table above
     * are used.  If any other NID reaches this function, there's a grave
     * coding error further down.
     */
    assert("Code that never should be reached" == NULL);
    return -1;
}
```

... and let `get_cipher_data_index` be used as before, and use `find_cipher_data_index` in that callback of yours.

Reply:
I can move that assertion to `get_cipher_method` and `destroy_cipher_method` where the NID has to be found.

Reply:
> ... and let get_cipher_data_index be used as before, and use find_cipher_data_index in that callback of yours.
I hadn't seen this comment.  I'll do that.